metricbeat/module/system/cpu/metrics/metrics.go

```diff
@@ -0,0 +1,249 @@
+// Licensed to Elasticsearch B.V. under one or more contributor
+// license agreements. See the NOTICE file distributed with
+// this work for additional information regarding copyright
+// ownership. Elasticsearch B.V. licenses this file to you under
+// the Apache License, Version 2.0 (the "License"); you may
+// not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing,
+// software distributed under the License is distributed on an
+// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+// KIND, either express or implied.  See the License for the
+// specific language governing permissions and limitations
+// under the License.
+
+package metrics
+
+import (
+	"fmt"
+	"reflect"
+
+	"github.com/pkg/errors"
+
+	"github.com/elastic/beats/v7/libbeat/common"
+)
+
+// CPU manages the CPU metrics from /proc/stat
+// If a given metric isn't available on a given platform,
+// The value will be null. All methods that use these fields
+// should assume that any value can be null.
+// The values are in "ticks", which translates to milliseconds of CPU time
+type CPU struct {
+	user    *uint64 `metric:"user"`
```

Comment:
Do we need all these fields to be optional? User/Sys seems to be available always, right?

I'm ok with using pointers. For completenes alternatives to pointers might be:

1. Introduce an 'OptUint' type:

```
type OptUint struct {
  isSet bool
  value uint64
}

func (opt OptUint) IsSet() bool { return opt.isSet }
func (opt *OptUint) Set(v uint64) { opts.isSet = true; opt.value = v }
func (opt *OptUint) Clear() { opt.isSet = false; }
func (opt *OptUint) Take() (uint64, bool) { ... } 

type CPU struct {
  user OptUint 
  sys OptUint
  ...
}
```

2. Use flags to describe 'capabilities' (supported fields):

```
type CPUFieldCap uint64

const (
  CPUFieldCapUser CPUFieldCap  = 1 << iota
  CPUFieldCapSys
  ...
)

type CPU struct {
  Fields CPUFieldCap

  user uint64
  ...
}
```

3. Use pointers, but with backing array (if you don't like to allocate):

```
type CPU struct {
  sys *uint64
  ...

  values [8]uint64
}
```

Reply:
Yah, your first suggestion is the one I've been contemplating the most, as it's the least awkward.